src/main/java/org/sonatype/repository/vgo/internal/util/VgoDataAccess.java

```diff
@@ -135,4 +147,40 @@ public Content toContent(final Asset asset, final Blob blob) {
     Content.extractFromAsset(asset, HASH_ALGORITHMS, content.getAttributes());
     return content;
   }
+
+  @TransactionalStoreBlob
+  public Content doCreateOrSaveComponent(final Repository repository,
+                                         final VgoAttributes vgoAttributes,
+                                         final String assetPath,
+                                         final TempBlob componentContent,
+                                         final Payload payload,
+                                         final VgoAssetKind assetKind) throws IOException
+  {
+    StorageTx tx = UnitOfWork.currentTx();
+    Bucket bucket = tx.findBucket(repository);
+
+    Component component = findComponent(tx,
+        repository,
+        vgoAttributes.getModule(),
+        vgoAttributes.getVersion());
+
+    if (component == null) {
+      component = tx.createComponent(bucket, repository.getFormat())
+          .name(vgoAttributes.getModule())
+          .version(vgoAttributes.getVersion());
+    }
+    tx.saveComponent(component);
```

Comment:
component only needs to be saved if it doesn't already exist.

Reply:
https://github.com/sonatype-nexus-community/nexus-repository-vgo/pull/5/commits/3d9cb3ab3daff08e2e79b0c1ba1002af9577992c